src/openvic-simulation/diplomacy/CountryRelation.hpp

```diff
@@ -46,50 +48,180 @@ namespace OpenVic {
+		using relation_value_type = int16_t;
+		class influence_value_type {
+			static constexpr uint8_t MAX_VALUE = 100;
+			uint8_t value;
+
+		public:
+			constexpr influence_value_type() = default;
+			constexpr influence_value_type(uint8_t value) : value(std::min(value, MAX_VALUE)) {}
+
+#define BIN_OPERATOR(OP) \
+	template<typename T> \
+	friend inline constexpr influence_value_type operator OP(influence_value_type const& lhs, T const& rhs) { \
+		return std::min<uint8_t>(lhs.value OP rhs, MAX_VALUE); \
+	} \
+	template<typename T> \
+	friend inline constexpr influence_value_type operator OP(T const& lhs, influence_value_type const& rhs) { \
+		return std::min<uint8_t>(lhs OP rhs.value, MAX_VALUE); \
+	}
+
+			BIN_OPERATOR(+);
+			BIN_OPERATOR(-);
+			BIN_OPERATOR(*);
+			BIN_OPERATOR(/);
+			BIN_OPERATOR(%);
+
+#undef BIN_OPERATOR
+
+			template<typename T>
+			friend inline constexpr bool operator==(influence_value_type const& lhs, T const& rhs) {
+				return lhs.value == rhs;
+			}
+			template<typename T>
+			friend inline constexpr auto operator<=>(influence_value_type const& lhs, T const& rhs) {
+				return lhs.value <=> rhs;
+			}
+
+
+#define BIN_OPERATOR_EQ(OP) \
+	template<typename T> \
+	friend inline constexpr influence_value_type& operator OP##=(influence_value_type & lhs, T const& rhs) { \
+		lhs.value = std::min<uint8_t>(lhs.value OP rhs, MAX_VALUE); \
+		return lhs; \
+	} \
+	friend inline constexpr influence_value_type& operator OP##=( \
+		influence_value_type & lhs, influence_value_type const& rhs \
+	) { \
+		return lhs OP## = rhs.value; \
+	}
+
+			BIN_OPERATOR_EQ(+);
+			BIN_OPERATOR_EQ(-);
+			BIN_OPERATOR_EQ(*);
+			BIN_OPERATOR_EQ(/);
+			BIN_OPERATOR_EQ(%);
+
+#undef BIN_OPERATOR_EQ
+		};
+
+		enum class OpinionType : int8_t {
+			Hostile = -2, //
+			Opposed,
+			Neutral,
+			Cordial,
+			Friendly,
+			Sphere
+		};
+
 	private:
-		// TODO: reference of manager responsible for storing CountryInstances
-		ordered_map<CountryRelationPair, country_relation_value_t> country_relations;
-
-	public:
-		CountryRelationManager(/* TODO: Country Instance Manager Reference */);
-
-		bool add_country(CountryRelationInstanceProxy country);
-		bool remove_country(CountryRelationInstanceProxy country);
-
-		country_relation_value_t
-		get_country_relation(CountryRelationInstanceProxy country, CountryRelationInstanceProxy recepient) const;
-		country_relation_value_t*
-		get_country_relation_ptr(CountryRelationInstanceProxy country, CountryRelationInstanceProxy recepient);
-		bool set_country_relation(
-			CountryRelationInstanceProxy country, CountryRelationInstanceProxy recepient, country_relation_value_t value
-		);
+#define RELATION_PAIR_MAP(PAIR_TYPE, VALUE_TYPE, NAME, FUNC_NAME, RECEPIENT_CONST) \
+	deque_ordered_map<PAIR_TYPE, VALUE_TYPE> NAME; \
+\
+public: \
+	VALUE_TYPE get_##FUNC_NAME(CountryInstance const* country, CountryInstance const* recepient) const; \
+	VALUE_TYPE& assign_or_get_##FUNC_NAME( \
```

Comment:
```suggestion
	VALUE_TYPE& get_##FUNC_NAME_reference( \
```
assign_or_get sounds very weird imo.

Reply:
Its not a get function, it assigns if it can't find it.